### PR TITLE
:tada: Source Pinterest - new streams

### LINKED ISSUE #22265: Source source-pinterest: implement AvailabilityStrategy
## What
We want GA connectors to make use of an AvailabilityStrategy in their streams to be able to perform error handling on the
stream level. The base use case supported by the `HttpAvailabilityStrategy` is for stream-based permissions,
in order to be able to skip a stream that the connector no longer has access to (because of e.g. scope
changes on an Oauth user config). An `AvailabilityStrategy` can be used to address any error that may occur
on a stream level, to avoid hitting errors, or having to except the same error for every record in a stream,
instead of skipping that stream altogether.

## How
There are a few different ways to go about this, usually depending on how much custom error handling we already
do in a connector, but also due to how the connector works.
- If we have no custom error handling: Inherit the default HttpAvailabilityStrategy. This is as simple as removing the override setting that currently sets
  the AvailabilityStrategy for the streams to `None`. This is useful if we don't currently do much custom error handling
  in the connector.
- If we have custom error handling for 403: Forbidden errors: Do the above to inherit the HttpAvailabilityStrategy,
  removing the custom error handling, but continuing any testing we have that 403s are handled gracefully by the connector
- If we have custom error handling for lots of per-stream errors: Extend the HttpAvailabilityStrategy. You can subclass
  the HttpAvailabilityStrategy to change the `reasons_for_unavailable_status_codes` method to except more error codes and
  provide messages for those error codes.
- If there is a smarter way to determine availability than trying to read the first record and excepting errors: Inherit
  the AvailabilityStrategy base class. For example, if there are clearly defined scopes required to access each stream in
  a given source, you might instead determine whether a stream is available by checking the scopes on the authenticator to
  see if it has the scopes required by the stream.

Note that the AvailabilityStrategy is defined on each stream. For most cases, it is easiest to set it in your stream's
base stream class, where the override is, but for some use cases, different streams might have different availability strategies -
and that's totally fine.

### PR DESCRIPTION
**Pull request by midavadim**
## What
Add missing streams
MLP: https://docs.google.com/document/d/1QfrLntfjJ-mKJHcn2O8eh9e46I6qhcjXtbD9vJmzQ98/edit

PIN_PROMOTION
PIN_PROMOTION_REPORT
PIN_PROMOTION_TARGETING_REPORT
PRODUCT_GROUP_REPORT
PRODUCT_GROUP_TARGETING_REPORT
KEYWORD
KEYWORD_REPORT
PRODUCT_ITEM_REPORT
CATALOG_FEED
CATALOG_PRODUCT_GROUPS
CUSTOMER_LIST
CONVERSION_TAG
AUDIENCE

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*
no impact

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/resources/pull-requests-handbook)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:generateScaffolds` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-pinterest/source_pinterest/streams.py [hunk 1]
```diff
@@ -137,6 +137,37 @@ def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
...
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return f"ad_accounts/{stream_slice['parent']['id']}/audiences"
+
+
+class Keywords(PinterestSubStream, PinterestStream):
+    """Docs: https://developers.pinterest.com/docs/api/v5/#operation/keywords/get"""
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        # return f"ad_accounts/{stream_slice['parent']['id']}/keywords?ad_group_id=2680068678993"
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-pinterest/source_pinterest/streams.py:151]: Please get rid of the commented lines

**Review by bazarnov — APPROVED** [review]

**Comment by alafanechere** [comment]
@midavadim I pulled in master to make the CI format check run correctly

File: airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml [hunk 2]
```diff
@@ -23,26 +23,24 @@ acceptance_tests:
...
+          - name: catalogs_product_groups
+            bypass_reason: Not possible to add data
+          - name: product_group_report
+            bypass_reason: Not possible to add data
+          - name: product_group_targeting_report
+            bypass_reason: Not possible to add data
+          - name: product_item_report
+            bypass_reason: Not possible to add data
+          - name: keyword_report
+            bypass_reason: Not possible to add data
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml:43]: `Not possible to add data`: does it mean the stream is empty or that is has default data we can't change?
If it's completely empty should we just pass a custom `configured_catalog_path` excluding these streams?

File: airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml [hunk 3]
```diff
@@ -54,6 +52,33 @@ acceptance_tests:
...
+              bypass_reason: "because it contains non-secure http link, which leads to failed QA tests"
+            - name: "media"
+              bypass_reason: "because it contains non-secure http link, which leads to failed QA tests"
+          board_section_pins:
+            - name: "links"
+              bypass_reason: "because it contains non-secure http link, which leads to failed QA tests"
+            - name: "link"
+              bypass_reason: "because it contains non-secure http link, which leads to failed QA tests"
+            - name: "media"
+              bypass_reason: "because it contains non-secure http link, which leads to failed QA tests"
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml:65]: We should rather make the QA check tolerate http links in expected records files.

**Review by alafanechere — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml (hunk 3 above)
Comment (alafanechere) [reply airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml:65]: @midavadim please review https://github.com/airbytehq/airbyte/pull/32499

File: airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml (hunk 2 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml:43]: it is empty streams - with no data.
As far as I know "Basic Test" does not allow to provide custom catalog

File: airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml (hunk 3 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml:65]: @alafanechere it would be great to have that PR merged!
Thank you!

Comment (alafanechere) [reply airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml:65]: It is merged, please remove these bypass 🙏 
I think you also have to increase the unit test coverage to get a 🟢 CI

File: airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml [hunk 4]
```diff
@@ -54,6 +52,33 @@ acceptance_tests:
...
+          board_section_pins:
+            - name: "links"
+              bypass_reason: "because it contains non-secure http link, which leads to failed QA tests"
+            - name: "link"
+              bypass_reason: "because it contains non-secure http link, which leads to failed QA tests"
+            - name: "media"
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-pinterest/acceptance-test-config.yml:64]: The QA checks should now pass. Please remove these bypass.

**Review by alafanechere — CHANGES_REQUESTED** [review]

**Review by alafanechere — APPROVED** [review]
Thanks for the fix, please make sure to have a 🟢 Connectors CI before merging

**Pull request merged**